### PR TITLE
[Bug Fix] NF Advanced Ring Thread Process NF Shutdown Messages

### PR DESCRIPTION
**Pull request by dennisafa**
When running the advanced rings mode in speed tester and scaling, closing the manager would not kill the NF running. (referencing #105 )

## Summary:
This adds a message handler in the advanced rings loop that listens for the MSG_STOP macro sent by the manager. 

**Usage:**

| This PR includes         |          |
| ------------------------ | -------- |
| Resolves issues          |  
| Breaking API changes     |
| Internal API changes     |
| Usability improvements   |  
| Bug fixes                | 💥
| New functionality        |
| New NF/onvm_mgr args     | 
| Changes to starting NFs  |  
| Dependency updates       | 
| Web stats updates        | 

## Merging notes:
 - Dependencies: None  

**TODO before merging :**
 - [x] PR is ready for review

## Test Plan:
Run the manager, then run speed tester (or scaling) with advanced ring mode: `./go.sh 1 -d 1 -a` 
Close the manager, assure that the NF is stopping. 

## Review: 
@koolzz @kevindweb I was thinking for sanity checks maybe try running several of the same NF and speed tester + scaling then closing manager.

* Sanity checks, assigned to @koolzz @kevindweb
  * Run make in /onvm and /examples directories
  * Test new functionality or bug fix from the pr (if needed)

* Code style, assigned to @koolzz @kevindweb
  * Run linter
  * Check everything style related

* Code design, assigned to @koolzz @kevindweb

  * Check for memory leaks
  * Check that code is reusable
  * Code is clean, functions are concise
  * Verify that edge cases properly handled

* Documentation, assigned to @koolzz @kevindweb
  * Check if the new changes are well documented, in both code and READMEs

(optional) **Subscribers:** << @-mention people who probably care about these changes >>

### CODE REVIEW

**Comment by onvm** [comment]
> In response to PR creation

## CI Message

Your results will arrive shortly

**Comment by onvm** [comment]
> In response to PR creation

## CI Message

Run successful see results:
[Results from nimbnode30]
Median TX pps for Speed Tester: 35184761

## Linter Failed

examples/arp_response/arp_response.c:284:  Lines should be <= 120 characters long  [whitespace/line_length] [5]
Total errors found: 1
examples/flow_table/openflow.h:50:  Using deprecated casting style.  Use static_cast<int>(...) instead  [readability/casting] [4]
examples/flow_table/openflow.h:569:  Extra space before ( in function call  [whitespace/parens] [4]
examples/flow_table/openflow.h:634:  Extra space before ( in function call  [whitespace/parens] [4]
examples/flow_table/openflow.h:771:  Extra space before ( in function call  [whitespace/parens] [4]
examples/flow_table/openflow.h:804:  Extra space before ( in function call  [whitespace/parens] [4]
examples/flow_table/openflow.h:865:  Extra space before ( in function call  [whitespace/parens] [4]
examples/flow_table/openflow.h:926:  Extra space before ( in function call  [whitespace/parens] [4]
Total errors found: 7
examples/nf_router/nf_router.c:173:  Lines should be <= 120 characters long  [whitespace/line_length] [5]
Total errors found: 1
examples/payload_scan/payload_scan.c:95:  Lines should be <= 120 characters long  [whitespace/line_length] [5]
examples/payload_scan/payload_scan.c:114:  Almost always, snprintf is better than strcpy  [runtime/printf] [4]
examples/payload_scan/payload_scan.c:126:  Lines should be <= 120 characters long  [whitespace/line_length] [5]
Total errors found: 3
onvm/onvm_mgr/onvm_pkt.c:68:  Are you taking an address of a cast?  This is dangerous: could be a temp var.  Take the address before doing the cast, rather than after  [runtime/casting] [4]
Total errors found: 1
onvm/onvm_mgr/onvm_stats.c:306:  If an else has a brace on one side, it should have it on both  [readability/braces] [5]
Total errors found: 1
onvm/onvm_nflib/onvm_pkt_common.c:98:  Are you taking an address of a cast?  This is dangerous: could be a temp var.  Take the address before doing the cast, rather than after  [runtime/casting] [4]
Total errors found: 1

File: examples/scaling_example/scaling.c [hunk 1]
```diff
@@ -289,6 +291,16 @@ run_advanced_rings(struct onvm_nf_info *nf_info) {
         }
 
         while (keep_running && rx_ring && tx_ring && nf) {
+                /* Check for a stop message from the manager. */
+                msg_q = nf->msg_q;
+                if (likely(rte_ring_count(msg_q) > 0)) {
```
Comment (koolzz) [inline examples/scaling_example/scaling.c]: Did you mean unlikely?

Comment (koolzz) [inline examples/scaling_example/scaling.c]: Did you mean unlikely?

File: examples/scaling_example/scaling.c [hunk 2]
```diff
@@ -289,6 +291,16 @@ run_advanced_rings(struct onvm_nf_info *nf_info) {
...
+                        msg = NULL;
+                        rte_ring_dequeue(msg_q, (void **) (&msg));
+                        if (msg->msg_type == MSG_STOP) {
+                                keep_running = 0;
+                        }
```
Comment (koolzz) [inline examples/scaling_example/scaling.c]: else
print received msg %d, ignoring

File: examples/scaling_example/scaling.c [hunk 3]
```diff
@@ -289,6 +291,16 @@ run_advanced_rings(struct onvm_nf_info *nf_info) {
         }
 
         while (keep_running && rx_ring && tx_ring && nf) {
+                /* Check for a stop message from the manager. */
```
Comment (koolzz) [inline examples/scaling_example/scaling.c]: Put this after the packet processing, best to copy this order https://github.com/sdnfv/openNetVM-dev/blob/develop/onvm/onvm_nflib/onvm_nflib.c#L533

File: examples/scaling_example/scaling.c (hunk 2 above)
Comment (koolzz) [inline examples/scaling_example/scaling.c]: also don't memleak :)
rte_mempool_put(nf_msg_pool, (void *)msg);

File: examples/scaling_example/scaling.c [hunk 4]
```diff
@@ -289,6 +291,16 @@ run_advanced_rings(struct onvm_nf_info *nf_info) {
...
+                msg_q = nf->msg_q;
```
Comment (koolzz) [inline examples/scaling_example/scaling.c]: put this before the while loop

File: examples/speed_tester/speed_tester.c [hunk 5]
```diff
@@ -333,6 +335,16 @@ run_advanced_rings(struct onvm_nf_info *nf_info) {
         start_time = rte_get_tsc_cycles();
 
         while (keep_running && rx_ring && tx_ring && nf) {
+                /* Check for a stop message from the manager. */
```
Comment (koolzz) [inline examples/speed_tester/speed_tester.c]: you got the idea, all of the comments also apply to this one

File: examples/scaling_example/scaling.c [hunk 6]
```diff
@@ -289,6 +291,16 @@ run_advanced_rings(struct onvm_nf_info *nf_info) {
...
+                if (likely(rte_ring_count(msg_q) > 0)) {
+                        msg = NULL;
+                        rte_ring_dequeue(msg_q, (void **) (&msg));
```
Comment (koolzz) [inline examples/scaling_example/scaling.c]: looks like extra space between (void **) and (&msg)

File: examples/scaling_example/scaling.c (hunk 3 above)
Comment (dennisafa) [reply examples/scaling_example/scaling.c]: We have a line before the actual packet processing that makes the loop continue if the number of packets is 0. Therefore, if I put the message check after that it won't work. Should we keep this check?

File: examples/scaling_example/scaling.c (hunk 2 above)
Comment (dennisafa) [reply examples/scaling_example/scaling.c]: nf_msg_pool isn't defined here, so I believe I'll have to grab it from onvm_nflib..

File: examples/scaling_example/scaling.c [hunk 7]
```diff
@@ -289,6 +292,19 @@ run_advanced_rings(struct onvm_nf_info *nf_info) {
...
+                if (unlikely(rte_ring_count(msg_q) > 0)) {
+                        msg = NULL;
+                        rte_ring_dequeue(msg_q, (void **)(&msg));
+                        if (msg->msg_type == MSG_STOP) {
+                                keep_running = 0;
+                        }
+                        else {
+                                printf("Received message %d, ignoring", msg->msg_type);
+                        }
+                        rte_mempool_put(rte_mempool_lookup(_NF_MSG_POOL_NAME), (void *)msg);
```
Comment (koolzz) [inline examples/scaling_example/scaling.c]: Save the output of rte_mempool_lookup(_NF_MSG_POOL_NAME) to a variable, looking it up every time isn't a good coding practice

**Comment by koolzz** [comment]
@onvm I need a pr with linter conflicts

**Comment by onvm** [comment]
> @onvm I need a pr with linter conflicts

## CI Message

Your results will arrive shortly

**Comment by onvm** [comment]
> @onvm I need a pr with linter conflicts

## CI Message

Error: ERROR: Script failed on nimbnode30

**Comment by koolzz** [comment]
@onvm should be good to go

**Comment by onvm** [comment]
> @onvm should be good to go

## CI Message

Your results will arrive shortly

**Comment by koolzz** [comment]
@onvm  jk, but now for sure

**Comment by onvm** [comment]
> @onvm  jk, but now for sure

## CI Message

Your results will arrive shortly

**Comment by koolzz** [comment]
@onvm is the linter pleased

**Comment by onvm** [comment]
> @onvm is the linter pleased

## CI Message

Your results will arrive shortly

**Comment by dennisafa** [comment]
@onvm how's your weekend been?

**Comment by onvm** [comment]
> @onvm how's your weekend been?

## CI Message

Another CI run in progress, please try again in 15 minutes

File: examples/speed_tester/speed_tester.c [hunk 8]
```diff
@@ -334,6 +334,9 @@ run_advanced_rings(struct onvm_nf_info *nf_info) {
         tx_ring = nf->tx_q;
         msg_q = nf->msg_q;
 
+        /* Access the nf msg memory pool */
+        nf_msg_pool = rte_mempool_lookup(_NF_MSG_POOL_NAME);
+
```
Comment (koolzz) [inline examples/speed_tester/speed_tester.c]: For future reference, prefer not to add to much white space. A good rule of thumb is to separate by newlines logical statements. Basic variable assignment shouldn't be separated

**Pull request merged**